core/src/main/java/io/undertow/server/handlers/accesslog/DefaultAccessLogReceiver.java

```diff
@@ -153,75 +154,66 @@ public void logMessage(final String message) {
      */
     @Override
     public void run() {
+        //check if we can transition to 2. If so, perform tasks in small chunks and check this.closed.
+        //move into 3 if(this.closed) and terminate run()
         if (!stateUpdater.compareAndSet(this, 1, 2)) {
             return;
         }
-        if (forceLogRotation) {
-            doRotate();
+        //NOTE: once we are here, run() control state transition, unless it is too slow
+        //and close takes over after grace period.
+
+        if (forceLogRotation || System.currentTimeMillis() > changeOverPoint) {
+            performFileRotation();
         } else if (initialRun && Files.exists(defaultLogFile)) {
-            //if there is an existing log file check if it should be rotated
-            long lm = 0;
-            try {
-                lm = Files.getLastModifiedTime(defaultLogFile).toMillis();
-            } catch (IOException e) {
-                UndertowLogger.ROOT_LOGGER.errorRotatingAccessLog(e);
-            }
-            Calendar c = Calendar.getInstance();
-            c.setTimeInMillis(changeOverPoint);
-            c.add(Calendar.DATE, -1);
-            if (lm <= c.getTimeInMillis()) {
-                doRotate();
-            }
+            checkAndRotateOnInitialRun();
         }
-        initialRun = false;
-        List<String> messages = new ArrayList<>();
-        String msg;
-        //only grab at most 1000 messages at a time
-        for (int i = 0; i < 1000; ++i) {
-            msg = pendingMessages.poll();
-            if (msg == null) {
-                break;
+        if(closed) {
+            if (!stateUpdater.compareAndSet(this, 2, 3)) {
+                UndertowLogger.ROOT_LOGGER.accessLogWorkerFailureOnTransition();
             }
-            messages.add(msg);
+            return;
         }
+
+        //only grab at most 1000 messages at a time
         try {
-            if (!messages.isEmpty()) {
-                writeMessage(messages);
-            }
-        } finally {
-            // change this state to final state
-            stateUpdater.set(this, 3);
-            //check to see if there is still more messages
-            //if so then run this again
-            if (!pendingMessages.isEmpty() || forceLogRotation) {
-                if (stateUpdater.compareAndSet(this, 3, 1)) {
-                    logWriteExecutor.execute(this);
+            for (int i = 0; i < 1000 && !this.closed && !pendingMessages.isEmpty(); ++i) {
+                final String msg = pendingMessages.peek();
```

Comment:
I might be looking too much into details here, but why peeking and removing instead of just polling?

Reply:
in case there is some intermitent IO failure, it should not purge entries without writing them first. Though, in such case, some overflow mechanism should be in place as well?

Reply:
Also "i < 1000 && !this.closed &&" checking this in loop might not be best, should be enough to loop through and check at the end IMHO